File: iOS/Sources/Quiz/Quiz View/Models/QuizResponse.swift
import Foundation

struct QuizResponse {
    let quiz: Quiz
    var respondent: Respondent
    var createdDate: Date
    var submittedDate: Date?
    var itemResponses: [ItemResponse]
}

extension QuizResponse {
    struct Respondent {
        var name: String = ""
        var store: String = ""
    }
    
    enum ItemResponse {
        case selectedResponseItemResponse(SelectedResponseItemResponse)
        case textInputItemResponse(TextInputItemResponse)
        case listItemResponse(ListItemResponse)
        
        var id: String {
            switch self {
            case .selectedResponseItemResponse(let item): item.id
            case .textInputItemResponse(let item): item.id
            case .listItemResponse(let item): item.id
            }
        }
        
        var itemId: String {
            switch self {
            case .selectedResponseItemResponse(let item): item.itemId
            case .textInputItemResponse(let item): item.itemId
            case .listItemResponse(let item): item.itemId
            }
        }
        
        func `as`<T>(_ type: T.Type) -> T {
            if T.self == SelectedResponseItemResponse.self {
                guard case .selectedResponseItemResponse(let value) = self else {
                    fatalError()
                }
                
                return value as! T
            }
            
            if T.self == TextInputItemResponse.self {
                guard case .textInputItemResponse(let value) = self else {
                    fatalError()
                }
                
                return value as! T
            }
            
            if T.self == ListItemResponse.self {
                guard case .listItemResponse(let value) = self else {
                    fatalError()
                }
                
                return value as! T
            }
            
            preconditionFailure()
        }
    }
    
    struct SelectedResponseItemResponse {
        let id: String
        let itemId: String
        let itemKind: Quiz.ItemKind
        var data = Data()
        
        struct Data {
            var selectedOptions: [SelectedOption] = []
            
            struct SelectedOption {
                var id: String
            }
        }
    }
    
    struct TextInputItemResponse {
        let id: String
        let itemId: String
        let itemKind: Quiz.ItemKind
        var data = Data()
        
        struct Data {
            var value: String = ""
        }
    }
    
    struct ListItemResponse {
        let id: String
        let itemId: String
        let itemKind: Quiz.ItemKind
        var data: Data
        
        struct Data {
            var itemResponses: [ItemResponse]
        }
    }
}

File: WebApp/src/quiz-app/models/QuizResponse.ts
import { Quiz } from './Quiz'

export interface QuizResponse {
  id: string
  createdDate: Date
  submittedDate: Date
  quiz: Quiz
  itemResponses: QuizResponse.ItemResponse[]
  respondent: {
    name: string
    store: string
  }
}

export type QuizResponsePayload = ReplaceFields<QuizResponse, {
  createdDate: string,
  submittedDate: string
}>

export namespace QuizResponse {
  export type ItemResponse =
    | ListItemResponse
    | SelectedResponseItemResponse
    | TextInputItemResponse

  export interface ListItemResponse {
    id: string
    itemId: string
    itemKind: "listItem"
    data: {
      itemResponses: ItemResponse[]
    }
  }

  export interface SelectedResponseItemResponse {
    id: string
    itemId: string
    itemKind: "selectedResponseItem"
    data: {
      selectedOptions: {
        id: string
        value: string
      }[]
    }
  }

  export interface TextInputItemResponse {
    id: string
    itemId: string
    itemKind: "textInputItem"
    data: {
      value: string
    }
  }
}

// Helper

type ReplaceFields<T, Replacements extends Partial<Record<keyof T, unknown>>> = {
  [K in keyof T]: K extends keyof Replacements ? Replacements[K] : T[K]
}